Lecture Review: Major Restructuring of Advanced Python Programming Section, Task 2-3 (#218)
QuantEcon/lecture-python-programming: lectures/debugging.md

(OK, sometimes we still do that...)

> But once you start writing larger programs you'll need a better system to spot bugs and errors.

Reviewer: ```suggestion
But once you start writing larger programs you'll need a better system.
```

> You may even want to handle possible errors in your code before they are raised.

Reviewer: ```suggestion
You may also want to handle potential errors in your code as they occur.
```

> In this lecture, we will discuss how to debug our programs and handle potential errors.

Reviewer: ```suggestion
In this lecture, we will discuss how to debug our programs and improve error handling.
```

The full list of magics is [here](http://ipython.readthedocs.org/en/stable/interactive/magics.html).

## Handling Errors

```{index} single: Python; Handling Errors
```

> Errors are problems within a program that stops the program from running further.

Reviewer 2: "stops" -> "stop"

Is this true?  If they are caught in a try-except block, then execution continues?

The full list of magics is [here](http://ipython.readthedocs.org/en/stable/interactive/magics.html).

> ## Handling Errors

Reviewer: I wonder if `Catching Errors` is a better title here as it is closer to the `try/catch` terminology in `python`

Author: Thanks for the comment @mmcky. I think this section also includes syntax errors and assertions. Maybe handling errors sounds like it is covering more than `try-catch`? But please correct me if you think `Catching Errors` is still a better title.

Reviewer: I agree let's keep `Handling Errors`

```{index} single: Python; Handling Errors
```

Sometimes it's possible to anticipate bugs and errors as we're writing code.

For example, the unbiased sample variance of sample $y_1, \ldots, y_n$
is defined as

$$
s^2 := \frac{1}{n-1} \sum_{i=1}^n (y_i - \bar y)^2
\qquad \bar y = \text{ sample mean}
$$

This can be calculated in NumPy using `np.var`.

But if you were writing a function to handle such a calculation, you might
anticipate a divide-by-zero error when the sample size is one.

One possible action is to do nothing --- the program will just crash, and spit out an error message.

But sometimes it's worth writing your code in a way that anticipates and deals with runtime errors that you think might arise.

Why?

* Because the debugging information provided by the interpreter is often less useful than the information

>   on possible errors you have in your head when writing code.

Reviewer: ```suggestion
* Because the debugging information provided by the interpreter is often less useful than what can be
  provided by a well written error message.
```

> * Because errors causing execution to stop are frustrating if you're in the middle of a large computation.

Reviewer: ```suggestion
* Because errors that cause execution to stop interrupt workflows
```

* supply specific information on why a program is failing

### Handling Errors During Runtime

```{index} single: Python; Runtime Errors
```

The approach used above is a bit limited, because it always leads to
termination.

Sometimes we can handle errors more gracefully, by treating special cases.

Let's look at how this is done.

#### Catching Exceptions

We can catch and deal with exceptions using `try` -- `except` blocks.

Here's a simple example

```{code-cell} python3
def f(x):
    try:
        return 1.0 / x
    except ZeroDivisionError:
        print('Error: division by zero.  Returned None')
    return None
```

When we call `f` we get the following output

```{code-cell} python3
f(2)
```

```{code-cell} python3
f(0)
```

```{code-cell} python3
f(0.0)
```

The error is caught and execution of the program is not terminated.

Note that other error types are not caught.

If we are worried the user might pass in a string, we can catch that error too

```{code-cell} python3
def f(x):
    try:
        return 1.0 / x
    except ZeroDivisionError:
        print('Error: Division by zero.  Returned None')
    except TypeError:

>         print('Error: Unsupported operation.  Returned None')

Reviewer: ```suggestion
        print(f'Error: x cannot be of type {type(x)}.  Returned None')
```

    return None
```

Here's what happens

```{code-cell} python3
f(2)
```

```{code-cell} python3
f(0)
```

```{code-cell} python3
f('foo')
```

If we feel lazy we can catch these errors together

```{code-cell} python3
def f(x):
    try:
        return 1.0 / x
    except (TypeError, ZeroDivisionError):
        print('Error: Unsupported operation.  Returned None')
    return None

> ```

Reviewer: Not sure this is needed. I suggest removing this as it reduces the clarity of the error message

Here's what happens

```{code-cell} python3
f(2)
```

```{code-cell} python3
f(0)
```

```{code-cell} python3
f('foo')
```

> If we feel extra lazy we can catch all error types as follows

Reviewer: ```suggestion
If you want to catch all error types you can also use:
```

```{code-cell} python3
def f(x):
    try:
        return 1.0 / x
    except:

>         print('Error.  Returned None')

Reviewer: ```suggestion
        print(f'Error.  An issue has occurred with x = {x} of type: {type(x)}')
```
